### ISSUE TITLE
All Types Missing in RTK Query

### ISSUE THREAD

**Issue by ianpaschal** [reduxjs/redux-toolkit#2667]
Hello,

I've been having some very frustrating issues with RTK for a while now with types not showing up. Below are some screen shots of what I'm seeing.

<img width="383" alt="Screen Shot 2022-09-02 at 12 36 14" src="https://user-images.githubusercontent.com/1137232/188123181-e67d8ffd-7d40-4dc9-a7f4-0f6b44fcdf39.png">

As you can see, `createApi()` does not specify return types. Neither do the hooks.

<img width="552" alt="Screen Shot 2022-09-02 at 12 44 36" src="https://user-images.githubusercontent.com/1137232/188123322-eeca6110-7a31-47ef-9d1b-63ac2dd3131f.png">

<img width="295" alt="Screen Shot 2022-09-02 at 12 36 22" src="https://user-images.githubusercontent.com/1137232/188123338-d18061c1-c94e-4ca5-b23a-d4a7ec204af4.png">

I could try to create a repo for it but this is a work project so it's quite a lot of work to strip it down to the point where I can share it, so I want to ask first if you have some ideas of common reasons this might occur.

As I said, I've had issues with RTK and types for quite a while now. I do have a reproduction repo for a similar issue, albeit with `configureStore()`. I suspect they're related, but if it's OK with you I'd rather get your initial diagnosis of the `createApi()` issue. Maybe I can demonstrate both issues in the same repo eventually.

Thanks in advance!

**Comment by phryneas** [comment]
That sounds like your TSServer is crashing for some reasons. It generally happens to me in bigger problems (with or without RTK) and restarting it usually makes the problem often go away. 

But unless you can reproduce that in a way it's pretty impossible to give any feedback (though this is probably more an issue that should be opened over at the TypeScript repo than here).

**Comment by ianpaschal** [comment]
Hmm, I didn't initially think TS as typing seems to work for all other packages in use. I figured maybe something might be off in my TS Config so I tried comparing (and all-out copying) [this one](https://github.com/rtk-incubator/rtk-query/blob/main/tsconfig.json) with no luck.

In this case, here's my [repro-repo](https://github.com/ianpaschal/rtk-type-issues). Admittedly it's a slightly different issue but I'll see if I can't get both to occur as I _bet_ they're related.

<img width="542" alt="Screen Shot 2022-09-02 at 19 22 54" src="https://user-images.githubusercontent.com/1137232/188206513-0a1e1366-f660-46cb-b26b-3f14944d08a8.png">

**Comment by markerikson** [comment]
@ianpaschal Ah. The _first_ issue is **you should _never_ add a `: Store` type!**

That's completely throwing away_ all_ the TS type inference that `configureStore` does.

Try removing that, and just have `const store = configureStore(/* etc*/)`.

**Comment by ianpaschal** [comment]
Aha! Indeed...

The good news is that solves the issue in my repro-repo.

The bad news is, that was one of my attempts to fix the problem. In my other project I didn't do that and the issue is still there.

<img width="786" alt="Screen Shot 2022-09-02 at 19 34 03" src="https://user-images.githubusercontent.com/1137232/188207692-76b98447-1609-4a0c-97ee-9bdc59af7b16.png">


I'll adapt the repro-repo to match this other project as closely as possible and see if I can reproduce it.

Thanks for your help though!

**Comment by markerikson** [comment]
Yeah, first thing I'd check is to see what the actual type of `store` is in in that case, before seeing what `RootState` is.

**Comment by ianpaschal** [comment]
Also `any` I'm afraid.

**Comment by ianpaschal** [comment]
Actually it's 7:40 PM here in Europe on a Friday so I'm going to put this down until Monday but I'll strip down this project until it's only RTK and push the changes to that other repository.

Again, thanks for your help! (and so quick too!)

**Comment by markerikson** [comment]
Yeah, at that point something's definitely wrong somewhere :)

**Comment by ianpaschal** [comment]
[Here's a reproduction repro](https://github.com/ianpaschal/redux-repro)

![Screen Shot 2022-09-07 at 11 53 17](https://user-images.githubusercontent.com/1137232/188849666-4064f4a4-e1f0-4de3-bee6-f1a36f635d53.png)
![Screen Shot 2022-09-07 at 11 53 04](https://user-images.githubusercontent.com/1137232/188849677-3fa768b2-8ce6-4046-8bb2-801f7a82ed7a.png)
 
 😞

**Comment by ianpaschal** [comment]
unfortunately this seems to also be preventing me from using `injectEndpoints` because if `internalApi` is any, then `build` is also any, and thus: `Untyped function calls may not accept type arguments.ts(2347)`

**Comment by phryneas** [comment]
Does it make a difference if you set strict: true in your tsconfig? RTK really should not be used without that.

**Comment by phryneas** [comment]
Apart from that, you might create a circular type here by importing store into the file. Generally you should not access the store in a `query` function - that one should be pure.

**Comment by ianpaschal** [comment]
Unfortunately `"strict": true` only means TypeScript now complains about `store`:

```
'store' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.ts(7022)
```

>Apart from that, you might create a circular type here by importing store into the file. Generally you should not access the store in a query function - that one should be pure.

...but that seems to be the issue! I feel a bit silly that didn't occur to me but on the other hand it's a bit of a bummer delegate that work to a React component (picking data out of Redux and then passing it back in RTK Query when that's global data unrelated to the component)... Obviously I can understand the appeal of keeping queries pure components but it would be nice to be able to be able to access the store directly. I guess the reason I've run into this several times is because I end up wanting some store data in my requests a lot of the time. User IDs, organization ID, etc... Food for thought I guess.

Anyway, that's this issue solved then!

Thank you for the help!

**Comment by phryneas** [comment]
> Unfortunately `"strict": true` only means TypeScript now complains about `store`:
> 
> ```
> 'store' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.ts(7022)
> ```

That means "you have a circular import there" - with strict, TS is warning you, without it it's just any.

> ...but that seems to be the issue! I feel a bit silly that didn't occur to me but on the other hand it's a bit of a bummer to have to delegate that work to a React component (picking data out of Redux and then passing it back in RTK Query when that's global data unrelated to the component)... Obviously I can understand the appeal of keeping queries pure components but it would be nice to be able to be able to access the store directly. I guess the reason I've run into this several times is because I end up wanting some store data in my requests a lot of the time. User IDs, organization ID, etc... Food for thought I guess.


The thing is: if that store data changes later, RTKQ doesn't know your query relied on it and cannot make a new request for new data. That is only possible if that data cones through `arg` since we can detect when that changes. If you get data from the side, you risk dispatching stale or wrong data.

**Comment by ianpaschal** [comment]
For posterity, when I made this thread I could have sworn I had asked about this before. And I had, back in July, but somehow couldn't find it back. I was pretty frazzled this summer planning my wedding and navigating the hell that is the Netherlands' housing market.

Anyway, just for the sake of finding things back: This is a duplicate of #2529. 🤦🏻
